### ISSUE TITLE
Deploy ECS ApplicationLoadBalancedFargateService isn't loading Secrets

### ISSUE THREAD

**Issue by josechifflet** [aws/aws-cdk#20744]
When inejcting secrets to the Secrets object of the `ApplicationLoadBalancedFargateService`'s `taskImageOptions` imported from 
`sm.Secret.fromSecretCompleteArn` (including 6 digit hypen), the deploy gets stuck and does not work.

I have also tried importing the secrets by:

`sm.Secret.fromSecretNameV2` and the behaviour is the same.

### Expected Behavior

ECS to deploy with custom secrets

### Current Behavior

CDK deploy freezed stuck at ECS apiService deploy

### Reproduction Steps

```
const importedSecrets = SECRET_NAMES.reduce((acc, key) => {
    const secret = sm.Secret.fromSecretCompleteArn(
      scope,
      `${config.ENVIRONMENT}/${key}`,
      secretFullArns[key], // full ARN of the secret identical as AWS
    );
    return {...acc, [key]: secret};
  }, {} as {[key: string]: sm.ISecret});
```
Where `SECRET_NAMES` is an array of secret names
Then,

```
  const secretsJson = SECRET_NAMES.reduce(
    (acc, key) => ({
      ...acc,
      [key]: ecs.Secret.fromSecretsManager(importedSecrets[key]),
    }),
    {},
  );
```

and then in the fargate service creator:

```
// Fargate service
this.backendService = new ecsPatterns.ApplicationLoadBalancedFargateService(
  scope,
  'apiService',
  {
    serviceName: 'apiService',
    cluster: this.ecsCluster,
    taskSubnets: {
      subnetType: ec2.SubnetType.PRIVATE_WITH_NAT,
    },
    memoryLimitMiB: 2048,
    cpu: 1024,
    desiredCount: 1,
    taskImageOptions: {
      containerName: 'apiContainer',
      image: ecs.ContainerImage.fromAsset('../api-backend/', {
        followSymlinks: SymlinkFollowMode.ALWAYS,
      }),
      containerPort: config.PORT,
      environment: {
        NODE_ENV: 'development',
        AWS_DEFAULT_REGION: config.AWS_DEFAULT_REGION,
      },
      logDriver: ecs.LogDrivers.awsLogs({
        streamPrefix: `${config.PROJECT_NAME}-logStream`,
        logGroup: fargateLog,
      }),
      secrets: {
        DB_BUSINESS_PASSWORD: ecs.Secret.fromSecretsManager(
          dbBusinessPasswordSecret,
          'password',
        ),
        DB_BUSINESS_HOST: ecs.Secret.fromSecretsManager(
          dbBusinessPasswordSecret,
          'host',
        ),


        ...secretsJson. 
        // ===========>>> If i comment this line here, the deploy is successfull, but without the secrets i want.
        


      },
    },
  },
);
```

In the CDK template i see the correct secret name BUT WITHOUT THE HYPHEN

CDK TEMPLATE Secrets from ECS:

```
....
{
  "Name": "X_API_KEY",
  "ValueFrom": {
    "Fn::Join": [
      "",
      [
        "arn:",
        {
          "Ref": "AWS::Partition"
        },
        "secret-arn-as-it-is-in-aws-console/co****/staging/X_API_KEY"
// The secret arn is exactly equal as it is in aws but without the 6 char hyphen auto-generated when the secret was uploaded.
      ]
    ]
  }
},
...
```
Parameter name is: `/co****/staging/X_API_KEY`
But in the aws console the arn of the secret is exactly the same as the one in the cdk template but without the hyphen.
Im pretty sure that's the cause of the deploy stuck freeze.

Also,
IAM Permissions from the image task are fine, for example for the secret who's name is:
`/co***/staging/CIPHER_SECRET-??????` where the ? represent the auto hyphen 6 digit

[![Console Synth CDK template][1]][1]

  [1]: https://i.stack.imgur.com/sjlxF.png

### CDK CLI Version

"aws-cdk-lib": "^2.27.0",

### Node.js Version

14.19

### OS

Macos Monterey

### Language

Typescript

### Language Version

"typescript": "^4.7.2"

### Other information

CDK JSON

```
{
  "app": "npx ts-node --prefer-ts-exts bin/CdkStarter.ts",
  "context": {
    "@aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId": true,
    "@aws-cdk/aws-rds:lowercaseDbIdentifier": true,
    "@aws-cdk/aws-efs:defaultEncryptionAtRest": true,
    "@aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021": false,
    "@aws-cdk/core:stackRelativeExports": false
  }
}
```

**Comment by josechifflet** [comment]
@madeline-k Any thoughts ?

**Comment by loutente** [comment]
Same issue

**Comment by sp0ker** [comment]
I am uncertain if it's help guys, but for my ALBFargateService I used secret like that :

```python
# Secret - manually created
secret_catalogv2back = aws_secretsmanager.Secret.from_secret_attributes(self, id="catalogv2back-secret", secret_complete_arn="arn:aws:secretsmanager:XXXXXX:XXXXXXXX:secret:catalogv2back-XXXXX")
# this return the ISecret object needed
```

And in the ALBFargateService
```python
service_catalogv2back = ecs_patterns.ApplicationLoadBalancedFargateService(self, id="catalogv2back",
            # options...
            task_image_options=ecs_patterns.ApplicationLoadBalancedTaskImageOptions(
                # options...
                secrets={
                    "APP_PORT": ecs.Secret.from_secrets_manager(secret_catalogv2back, "APP_PORT"),
                    "DATABASE_HOST": ecs.Secret.from_secrets_manager(db_catalogv2back.secret, "host"), # ISecret from the ServerlessCluster object
                },
```

Regards,

**Comment by mrgrain** [comment]
Hi @josechifflet and thanks for reporting this issue.

Could you please clarify what you mean by "CDK deploy freezed stuck at ECS apiService deploy"?
Are you receiving any error messages?
How far does the deployment progress before the described behavior occurs?

**Comment by bryan-queryai** [comment]
@mrgrain I just ran into this problem as well and the short of it is that when importing a secret only `Secret.from_secret_complete_arn` results in a correct iam policy attached to the role and a correct entry in the task definition.  Both .from_secret_partial_arn and .from_secret_name_v2 methods fail to include the arn's logical id and thus don't produce proper iam policy or ecs task definitions.

Updating the docs to explain which Secret import method to use is probably sufficient imo.

Here's some code snippets if it helps explain the issue better!

```python
from aws_cdk import aws_secretsmanager as secrets_manager
...

# using .from_secret_complete_arn works
foo_secret = secrets_manager.Secret.from_secret_complete_arn(
    scope=self,
    id="dev-foo-secret",
    secret_complete_arn="arn:aws:secretsmanager:us-east-1:1234567890:secret:dev-foo-secret-abCDEF"
)

# .from_secret_complete_arn creates a policy with the logical id
""" {
    "Action": [
        "secretsmanager:DescribeSecret",
        "secretsmanager:GetSecretValue"
    ],
    "Effect": "Allow",
    "Resource": "arn:aws:secretsmanager:us-east-1:1234567890:secret:dev-foo-secret-abCDEF"
} """

### these methods of importing a secret for use with ECS result in an IAM policy that lacks the logical id in the arn

# this method fails; I think this one should work and the resulting foo_secret.secret_full_arn should have the logical id, however, the policy doesn't include it
foo_secret = secrets_manager.Secret.from_secret_partial_arn(
    scope=self,
    id="dev-foo-secret",
    secret_partial_arn=f"arn:aws:secretsmanager:us-east-1:1234567890:secret:dev-foo-secret"
)

# this method fails; I think this one is expected to fail as importing from name is known to lack the logical id
foo_secret = secrets_manager.Secret.from_secret_name_v2(
    scope=self,
    id="dev-foo-secret",
    secret_name="dev-foo-secret"
)

# .from_secret_partial_arn and .from_secret_name_v2 result in a IAM policy attached to the Task Execution Role like this with `??????`
""" {
    "Action": [
        "secretsmanager:DescribeSecret",
        "secretsmanager:GetSecretValue"
    ],
    "Effect": "Allow",
    "Resource": "arn:aws:secretsmanager:us-east-1:1234567890:secret:dev-foo-secret-??????" # <--- missing logical id
} """

...

secrets={
    "FOO_SECRET": ecs.Secret.from_secrets_manager(self.foo_secret, "BAR")
},

# .from_secret_partial_arn and .from_secret_name_v2 result in the ECS Task def ValueFrom that doesn't have the `?????` but is still invalid
"""
"Name": "FOO_SECRET",
"ValueFrom": "arn:aws:secretsmanager:us-east-1:1234567890:secret:dev-foo-secret:BAR::" # <--- missing logical id
"""
```

**Comment by mrgrain** [comment]
@bryan-queryai Thanks for your detailed report. This helps clearing things up!
I also agree that most of it could be solved by improving the docs. We very much welcome contributions in this area.

> \# using .from_secret_complete_arn works
\# .from_secret_complete_arn creates a policy with the logical id

Works as expected. Nice!

> \### these methods of importing a secret for use with ECS result in an IAM policy that lacks the logical id in the arn
\# this method fails; I think this one should work and the resulting foo_secret.secret_full_arn should have the logical id, 
\# this method fails; I think this one is expected to fail as importing from name is known to lack the logical id

I think this is also expected. Specifically `from_secret_partial_arn` is referencing the secret without the suffix (hyphen and six character). Unfortunately CDK and CFN have no means of finding the correct suffix here.

>\# .from_secret_partial_arn and .from_secret_name_v2 result in a IAM policy attached to the Task Execution Role like this
>`"Resource": "arn:aws:secretsmanager:us-east-1:1234567890:secret:dev-foo-secret-??????" # <--- missing logical id`

This also works as expected. We don't know the suffix, so the policy is created with a wildcard matching any suffix.


> \# .from_secret_partial_arn and .from_secret_name_v2 result in the ECS Task def ValueFrom that doesn't have the `?????` but is still invalid
>`"ValueFrom": "arn:aws:secretsmanager:us-east-1:1234567890:secret:dev-foo-secret:BAR::" # <--- missing logical id`

This seems to be the core of the issue here. Although I'm not entirely sure yet if leaving the `::` would actually help much. 

Bottom line really is that `from_secret_complete_arn` is the only method that should be used in 99.9% of the use cases.

**Comment by mccauleyp** [comment]
I just spent a while being perplexed by this. I've been successfully using `Secret.from_secret_partial_arn()` together with `ApplicationLoadBalancedTaskImageOptions` and `aws_ecs.Secret.from_secrets_manager` for around 8 months in several projects, but now I'm getting an error that's resolved by using `Secret.from_secret_complete_arn()` instead. The CDK deployment unfortunately doesn't surface the error message, but I see the IAM Access Denied error in the ECS task console while the it tries to bring up the new task. 

What's most confusing is that I have two secrets, one created a couple months ago and one created today, and it's only the one created today that exhibits this problem requiring `from_secret_complete_arn`. I'm still using `.from_secret_partial_arn` for the old one. So that makes it seem like something is different about the secret, but I can't see any configuration difference. I created the plain-text secrets from the console using the default settings. 

Unfortunately using `from_secret_complete_arn` isn't quite as useful because we need to know the randomly-generated suffix. Right now our workflow is to create a small number of secrets with the same name manually in different environments, so then the CDK code references the same name in each environment. Now we'll need to make some changes to that to account for needing to know the suffix.

**Comment by dodge-m** [comment]
This still does not appear to be updated in the CDK API reference docs. Having the secret association require the full ARN of the secret seems to very inconvenient as now each secret ARN must be known ahead of time and have to be managed instead of allowing each env just just expect a secret with a particular name. Would be ideal if this could be fixed so that secret name importation could be used, but at minimum should have the docs tell you what will not work.

**Comment by pahud** [comment]
Hi @dodge-m, @mccauleyp, and others following this issue,

Thank you for the continued feedback. You're right that the documentation still needs improvement to clarify this behavior.

## The Core Issue

When importing secrets for use with ECS, you **must** use [`Secret.fromSecretCompleteArn()`](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-secretsmanager/lib/secret.ts#L653) with the full ARN including the 6-character suffix.

**Why other methods don't work with ECS:**
- `fromSecretNameV2()` and `fromSecretPartialArn()` don't know the 6-character suffix that AWS Secrets Manager automatically appends
- IAM policies can use wildcards (`-??????`), so they work fine
- ECS task definitions require the exact ARN in the `ValueFrom` field and don't support wildcards
- This mismatch causes the "Access Denied" errors you're seeing

## Working Example

```typescript
// ✅ This works with ECS
const secret = secretsmanager.Secret.fromSecretCompleteArn(
 this,
 'ImportedSecret',
 'arn:aws:secretsmanager:us-east-1:123456789012:secret:my-secret-abc123'
);

// ❌ These don't work with ECS (but work for other use cases)
const secret = secretsmanager.Secret.fromSecretNameV2(this, 'Secret', 'my-secret');
const secret = secretsmanager.Secret.fromSecretPartialArn(this, 'Secret', 'arn:...:secret:my-secret');
```
To get the complete ARN:
```bash
aws secretsmanager describe-secret --secret-id my-secret-name --query ARN --output text
```
## Next Steps

We acknowledge this needs better documentation. I'm labeling this as `documentation`. We need to update:

- `packages/aws-cdk-lib/aws-ecs/README.md` - Add ECS-specific secret import guidance
- `packages/aws-cdk-lib/aws-secretsmanager/README.md` - Clarify ECS compatibility for each import method
- JSDoc comments in `secret.ts` - Add warnings about ECS limitations

**Reference:** [AWS documentation on secret ARN formats](https://docs.aws.amazon.com/secretsmanager/latest/userguide/troubleshoot.html#ARN_secretnamehyphen)

I will submit an immediate doc update PR for that today.
